source/extensions/clusters/dns_srv/dns_srv_cluster.cc

```diff
@@ -0,0 +1,351 @@
+#include "source/extensions/clusters/dns_srv/dns_srv_cluster.h"
+
+#include <algorithm>
+#include <chrono>
+#include <list>
+#include <memory>
+#include <string>
+
+#include "envoy/common/exception.h"
+#include "envoy/config/cluster/v3/cluster.pb.h"
+#include "envoy/config/endpoint/v3/endpoint.pb.h"
+#include "envoy/stats/scope.h"
+
+#include "source/common/common/fmt.h"
+#include "source/common/network/utility.h"
+#include "source/common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Upstream {
+
+DnsSrvCluster::DnsSrvCluster(
+    const envoy::config::cluster::v3::Cluster& cluster,
+    const envoy::extensions::clusters::dns_srv::v3::DnsSrvClusterConfig& dns_srv_cluster,
+    ClusterFactoryContext& context, Network::DnsResolverSharedPtr dns_resolver,
+    absl::Status& creation_status)
+    : BaseDynamicClusterImpl(cluster, context, creation_status), cluster_(cluster),
+      dns_resolver_(dns_resolver),
+      dns_refresh_rate_ms_(
+          std::chrono::milliseconds(PROTOBUF_GET_MS_OR_DEFAULT(cluster, dns_refresh_rate, 5000))),
+      respect_dns_ttl_(cluster.respect_dns_ttl()),
+      resolve_timer_(context.serverFactoryContext().mainThreadDispatcher().createTimer(
+          [this]() -> void { startResolve(); })),
+      local_info_(context.serverFactoryContext().localInfo()), dns_srv_cluster_(dns_srv_cluster) {
+
+  dns_lookup_family_ = getDnsLookupFamilyFromCluster(cluster);
+
+  load_assignment_.add_endpoints()->add_lb_endpoints()->set_endpoint_name(
+      dns_srv_cluster_.srv_name());
+
+  ENVOY_LOG(debug, "starting async DNS resolution for {}", dns_srv_cluster_.srv_name());
+}
+
+void DnsSrvCluster::startPreInit() {
+  startResolve();
+  if (!wait_for_warm_on_init_) {
+    onPreInitComplete();
+  }
+}
+
+DnsSrvCluster::~DnsSrvCluster() {
+  if (active_dns_query_) {
+    active_dns_query_->cancel(Network::ActiveDnsQuery::CancelReason::QueryAbandoned);
+  }
+}
+
+void DnsSrvCluster::startResolve() {
+  ENVOY_LOG(debug, "starting async DNS resolution for {}", dns_srv_cluster_.srv_name());
+  info_->configUpdateStats().update_attempt_.inc();
+
+  active_resolve_list_.reset(new ResolveList(*this));
```

Comment:
Use `make_unique` when possible

```suggestion
  active_resolve_list_ = make_unique<ResolveList>(*this);
```